### ISSUE TITLE
Better name?

### ISSUE THREAD

**Issue by Danack** [Danack/RfcLiteralString#11]
Although the name 'literal' isn't terrible, can we come up with a better one?

compile-time constant is a phrase used in the Christoph Kern talk.....and while that is more accurate, it's not that pithy.

btw probably worth us thinking about it rather than asking elsewhere.....for reasons.

**Comment by craigfrancis** [comment]
Jakob Givoni suggested [is_from_literal()](https://news-web.php.net/php.internals/109197).

I can’t remember who, but someone suggested “is_safe_string()”, but I think that’s getting a bit vague, and I think Sara would have an issue with it.

The “compile time constants” description kinda works; but with the 2 functions to concatenate, it’s stretching that definition a bit.

I’ll also try to keep thinking of something else: where I quite like it following the “is_*” format for the other functions.

**Comment by Danack** [comment]
Maybe (apart from being longer):

* is_code_string()
* code_string_combine()
* code_string_implode()

might be nice. Yeah, no-one can guess what it means, but it seems quite easy to explain to a junior dev.

The only thing I prefer literal for, is that it would allow slipping in ints...

`literal_combine("foo=", intval($_GET['foo']));` is not an obvious wtf..

`code_string_combine("foo=", intval($_GET['foo']));` is quite odd.

**Comment by Danack** [comment]
On the other hand, if there isn't a better option, then 'literal string' has pretty clear search results.

https://www.google.com/search?q=what+is+literal+string+in+php

> A string literal is the notation for representing a string value within the text of a computer program. In PHP, strings can be created with single quotes, double quotes or using the heredoc or the nowdoc syntax. ... The heredoc preserves the line breaks and other whitespace (including indentation) in the text.


> What is an example of a string literal?
> A string literal is a sequence of zero or more characters enclosed within single quotation marks. The following are examples of string literals: 'Hello, world!' 'He said, "Take it or leave it."'

**Comment by craigfrancis** [comment]
That's a really good point, existing articles/discussions on literal strings defines it pretty well.

---

With the ability to slip in ints... tbh I wasn't thinking of supporting this, only for consistency reasons (sometimes I explain this as checking for "programmer defined strings"), where this might seem to be going against the concept:

    $sql = literal_combine('SELECT * FROM user WHERE id = ', intval($_GET['id']));

That said, I can't think of an example where this could be abused.

**Comment by Danack** [comment]
> I can't think of an example where this could be abused.

Yeah, I've been testing out the words "strings can accidentally represent multiple parameters when passed to an API (particularly over http or other string based api), but ints are only ever going to be one parameter".

**Comment by Danack** [comment]
Some words: 
https://github.com/Danack/RfcLiteralString/blob/master/rfc_words.md#why-the-name-is_literal-

**Comment by craigfrancis** [comment]
"Why the name 'is_literal()'" is a really good addition to the RFC, thanks for that, it's given me some confidence it's the right name.

---

As to the int's thing, I think you're right.

But I'm not sure I'd use that explanation. While it does apply for a lot of the issues, problematic values could still be seen as a single parameter; e.g. `javascript:alert(1)` being added to `<a href="?">`, where the context for that value is the issue.

I've been wondering, with the literal string (e.g. SQL), could there be an issue where the recipient (e.g. database) is given an integer that's larger than what it can accept (e.g. 64bit PHP to a 32bit database), or maybe negative value? In comparison, if that same value was provided as a parameter (known to contain user values) that might result in situation being handled better (e.g. error vs no results).

And when it comes to SQL, the database could get a little upset with caching, on the basis that the SQL changes every time (because the id isn't provided as a parameter).

I don't think either of those affect MySQL (e.g. [where its Prepared Statements cache is based on table structure](https://dev.mysql.com/doc/refman/8.0/en/statement-caching.html), and the [Query Cache has been removed](https://dev.mysql.com/doc/refman/5.7/en/query-cache.html)), but it could affect something else? maybe?

Also, does this work for a float/double?
